### PR TITLE
[mono] Remove domain lock

### PR DESCRIPTION
**Pull request by vargaz**
* Avoid double locking the loader+domain lock, its not needed.
* Use the mem manager lock and the loader lock in a few places instead of
  the domain lock.
* Use lock free code in mono_jit_runtime_invoke ().
* Avoid locking around mono_jit_info_table_add (), it does locking itself.

### CODE REVIEW

File: src/mono/mono/metadata/loader-internals.h [hunk 1]
```diff
@@ -98,10 +97,15 @@ struct _MonoMemoryManager {
 	// Currently unused, we take the domain lock instead
```
Comment (lambdageek) [inline src/mono/mono/metadata/loader-internals.h:97]: This comment is wrong now.

```suggestion
	// Must be acquired after the loader lock
```

File: src/mono/mono/metadata/lock-tracer.h [hunk 2]
```diff
@@ -15,7 +15,6 @@ typedef enum {
 	InvalidLock = 0,
 	LoaderLock,
 	ImageDataLock,
-	DomainLock,
```
Comment (lambdageek) [inline src/mono/mono/metadata/lock-tracer.h:18]: I would like to see `MemoryManagedLock` in here.  And also a dotnet/runtime (.NET 6+) update to https://www.mono-project.com/docs/advanced/runtime/docs/thread-safety/#the-locking-hierarchy

File: src/mono/mono/metadata/memory-manager.c [hunk 3]
```diff
@@ -217,68 +217,77 @@ mono_mem_manager_free_singleton (MonoSingletonMemoryManager *memory_manager, gbo
...
-	mono_domain_lock (memory_manager->domain);
+	mono_coop_mutex_lock (&memory_manager->lock);
 }
 
 void
 mono_mem_manager_unlock (MonoMemoryManager *memory_manager)
 {
-	//mono_coop_mutex_unlock (&memory_manager->lock);
-	mono_domain_unlock (memory_manager->domain);
+	mono_coop_mutex_unlock (&memory_manager->lock);
```
Comment (lambdageek) [inline src/mono/mono/metadata/memory-manager.c:226]:
```suggestion
	mono_locks_coop_acquire (&memory_manager->lock, MemoryManagerLock);
```

**Review by lambdageek — APPROVED** [review]
we should update the lock hierarchy docs on mono-project.com.

might be nice to use the lock tracer stuff for the memory manager coop lock.

some minor nits for gslist.

File: src/mono/mono/metadata/object.c [hunk 4]
```diff
@@ -1626,7 +1626,9 @@ build_imt (MonoClass *klass, MonoVTable *vt, gpointer* imt, GSList *extra_interf
 {
 	MONO_REQ_GC_NEUTRAL_MODE;
 
+	mono_loader_lock ();
 	build_imt_slots (klass, vt, imt, extra_interfaces, -1);
```
Comment (lambdageek) [inline src/mono/mono/metadata/object.c:1630]: nit: AFAICT, `build_imt` is dead code - maybe we can just delete it.

**Comment by vargaz** [comment]
Will address comments in a follow up PR.

**Pull request merged**